### ISSUE TITLE
"Non-Deterministic" error which I have high confidence to believe cause by host rather than my code.

### ISSUE THREAD

**Issue by jjshao-ms** [Azure/azure-functions-durable-extension#2563]
### Description
> A clear and concise description of what the bug is. Please make an effort to fill in all the sections below; the information will help us investigate your issue.

Hi all, my orchestrator is getting "Non-Deterministic" errors which I believe isn't cause by my code.
The "Non-Deterministic" error message I got looks like this: `"Non-Deterministic workflow detected: A previous execution of this orchestration scheduled an activity task with sequence ID 0 and name 'xxxxx' (version ''), but the current replay execution hasn't (yet?) scheduled this task. Was a change made to the orchestrator code after this instance had already started running?"`
As you can see the sequence ID is _**"0"**_, which means it is the first activity that got orchestrated, but my orchestrator looks like following:
```csharp
[Function(OrchName)]
public async Task Run([OrchestrationTrigger] TaskOrchestrationContext context, FunctionContext functionContext)
{
    var data = await context.CallActivityAsync<List<string>>(nameof(this.MyActivity), null);

    if (!context.IsReplaying)
    {
        var log = functionContext.GetLogger(OrchName);
        log.LogInformation($"{data.Count} sources need to be updated.");
    }

    context.SetCustomStatus(new { count = data.Count });

    // more codes to process data
    foreach (var item in data) /* ... */
}
```
The first line of code my orchestrator was call the first activity, so there isn't any change my orchestrator didn't schedule a task with sequence ID "0". And I can guarantee that I didn't do a deploy, so the code didn't change.

### Additional information
Two things worth a mention:
Firstly, our orchestration never got "Non-Deterministic" error when we use in proc dotnet function, but just after we move to dotnet-isolated worker, it starts reporting this error.

Secondly, I did some code read of the durable function itself, I found some piece of code that might be the root cause of this:
When the executor replaying my orchestration, it doesn't execute my code directory, but firstly execute the isolated worker middleware, and there is an await operation exist!
https://github.com/Azure/azure-functions-dotnet-worker/blob/ce04c2794482dd97abf3a74e1f5ee5648d8591e1/src/DotNetWorker.Core/Invocation/DefaultFunctionExecutor.cs#L42
```csharp
// inside DefaultFunctionExecutor
inputBindingResult = await inputBindingFeature.BindFunctionInputAsync(context);
```
So, I guess this action was awaited asynchronously, so my code hasn't got a change to execute, resulting the first activity to be not scheduled and an "Non-Deterministic" was throwed!
Do you ever considered this situation? If you, how do you guarantee this await action will be completed synchronously?

### Expected behavior
> A clear and concise description of what you expected to happen.

My orchestrator shouldn't report "Non-Deterministic" error. At least, not on ID with "0".

### Actual behavior
> A clear and concise description of what actually happened.

"Non-Deterministic"

### Relevant source code snippets
see above.

### Known workarounds
> Provide a description of any known workarounds you used.

Hope you can provide one for me.

### App Details

- **Durable Functions extension version (e.g. v1.8.3)**:
```csproj
<PackageReference Include="Microsoft.Azure.Functions.Worker" Version="1.17.0" />
<PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.DurableTask" Version="1.0.2" />
<PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.Storage" Version="5.1.2" />
<PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.Timer" Version="4.2.0" />
<PackageReference Include="Microsoft.Azure.Functions.Worker.Sdk" Version="1.11.0" />
```
- **Azure Functions runtime version (1.0 or 2.0)**: 4.0
- **Programming language used**: C# (isolated)

### Screenshots
> If applicable, add screenshots to help explain your problem.

If you need any, tell me.

### If deployed to Azure
> We have access to a lot of telemetry that can help with investigations. Please provide as much of the following information as you can to help us investigate!

The latest one:
- **Timeframe issue observed**:  "createdTime": "2023-08-25T10:59:22Z", "lastUpdatedTime": "2023-08-26T15:43:24Z"
- **Function App name**: reach me through Teams if you need, search my GitHub user name
- **Function name(s)**: same as above
- **Azure region**: wus
- **Orchestration instance ID(s)**: "abaa0e2b34d15370a66898d6d3f625e9"
- **Azure storage account name**: same as above

Other old ones:
"createdTime": "2023-08-19T10:17:17Z"
"lastUpdatedTime": "2023-08-19T23:52:17Z"
"instanceId": "0079d620afca5ef3a531007facad5c12"

"createdTime": "2023-08-13T10:20:31Z"
"lastUpdatedTime": "2023-08-15T11:13:28Z"
"instanceId": "5e0251be87eb597c84a6d957f49c2e28"

> If you don't want to share your Function App or storage account name GitHub, please at least share the orchestration instance ID. Otherwise it's extremely difficult to look up information.

**Comment by davidmrdavid** [comment]
@jviau: are you aware of any non-determinism errors that may be caused by a failed cancellation flow?

**Comment by jviau** [comment]
Could it be related to this? https://github.com/microsoft/durabletask-dotnet/issues/158

**Comment by YanumVlad** [comment]
I am experiencing a similar issue with Python based durable function (containerized)

**Comment by jjshao-ms** [comment]
> Could it be related to this? [microsoft/durabletask-dotnet#158](https://github.com/microsoft/durabletask-dotnet/issues/158)

@jviau I don't think so, because our function does not have any custom middleware when this problem happens.
Though we do add a custom middleware this week, but that is just a log to help triage this problem, and it looks like following, I don't think it will cause Non-Deterministic, right?
```csharp
public Task Invoke(FunctionContext context, FunctionExecutionDelegate next)
{
    if (!IsOrchestrationTrigger(context))
    {
        return next(context);
    }

    this.logger.LogInformation("Orchestration start executing function: {functionId}, invocation: {invocationId}", context.FunctionId, context.InvocationId);
    return next(context);
}

private static bool IsOrchestrationTrigger(FunctionContext context)
{
    return context.FunctionDefinition.InputBindings.Values.Any(x => x.Type == "orchestrationTrigger");
}
```

**Comment by somatusjn** [comment]
I'm having the same issue and the code in my orchestrator is a simple as it gets: 

![Image](https://github.com/user-attachments/assets/2128e7a9-5725-45b2-80aa-eaacd7d19e26)

My function is having some memory issues, where the python code is exiting with code 137, and some runs show that error, but I'm not sure why some runs show this non-deterministic error.

**Comment by rafael-unloan** [comment]
Any updates on this @davidmrdavid ? (or whoever is now responsible for durable functions?)

**Comment by davidmrdavid** [comment]
@rafael-unloan - Although I'm no longer in the team, I did notice the linked issue (https://github.com/microsoft/durabletask-dotnet/issues/158) seems to have been resolved, so perhaps this will no longer be an issue in the next release (assuming they were related in the first place).

@jviau - any chance you could chime in here?
Also, I've added the "Needs: Triage" label here so this can be assigned to a proper DF team member. Thanks!
